base/server/python/pki/server/deployment/__init__.py

```diff
@@ -2575,6 +2575,11 @@ def setup_database_user(self, instance, subsystem):
             user_type='agentType',
             state='1')
 
+        logger.info('Applying paged size limit to pkidbuser')
+        subsystem.modify_user(
+            'pkidbuser',
+            add_see_also='nsPagedSizeLimit=20000')
+
```

Comment:
The `add_see_also` and `del_see_also` params were provided to add/remove a `seeAlso` attribute (which is multi-valued) in LDAP:
https://github.com/dogtagpki/pki/blob/master/base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java#L790-L793
https://github.com/dogtagpki/pki/blob/master/base/server/src/main/java/com/netscape/cmscore/usrgrp/UGSubsystem.java#L832-L835

To set an `nsPagedSizeLimit` attribute (which is single-valued) I'd suggest modifying `UGSubsystem` to do something like this:
```
LDAPModificationSet mods = new LDAPModificationSet();

if (seeAlsoToAdd != null) {
    LDAPAttribute attr = new LDAPAttribute("seeAlso");
    attr.addValue(seeAlsoToAdd);
    mods.add(LDAPModification.ADD, attr);
}

if (seeAlsoToRemove != null) {
    LDAPAttribute attr = new LDAPAttribute("seeAlso");
    attr.addValue(seeAlsoToRemove);
    mods.add(LDAPModification.DELETE, attr);
}

if (pagedSizeLimit != null) {
    LDAPAttribute attr = new LDAPAttribute("nsPagedSizeLimit");
    attr.addValue(pagedSizeLimit);
    mods.add(LDAPModification.REPLACE, attr);
}

ldapconn.modify(dn, mods);
```
So I think it will require some non-trivial changes, but it shouldn't be too complicated.

Reply:
Also, this code is currently not tested. You might want to modify an existing test to verify that the `nsPagedSizeLimit` attribute was indeed set in LDAP.

Reply:
Another thing is, this change will affect new installation only. The existing `pkidbuser` will not be updated and we currently don't have a mechanism to upgrade the database.
https://github.com/dogtagpki/pki/issues/1279
https://github.com/dogtagpki/pki/issues/1476

Also, since this issue is mainly affecting IPA, this upgrade probably can be done by IPA since IIUC they have a mechanism for that.

Reply:
> I'd suggest modifying `UGSubsystem` to do something like this

If the user is created in the UGSubSystem then the new size limit can be set during the creation of the user taking the value from the ldap configuration. The dynamic change based on the real size of the entries need some additional thought, maybe can be added to the pruning job which could have an estimation of the entries size.

Reply:
The current pruning job removes expired entries at a fixed rate regardless of database size. I'm not sure if DS provides a way to get the number of entries in the database. Maybe it's possible with VLV, but in the future we might drop VLV, so we shouldn't rely on that.

Reply:
If it is not possible to get the number of remaining entry a different approach is needed. That was just on option but we can create a separate process/thread/cron.

Reply:
Thanks for the feedback! I will take a look at modifying `UGSubsystem` and adding a test. I don't think the upgrade is necessary as the problem is easy to spot and the workaround is simple - just an LDAP modify command (which has been tested in the field by support).

Reply:
As this is going to simply add a static value, would it be reasonable to modify `addUser` to put `nsPagedSizeLimit` into the existing `LDAPAttributeSet`?

Reply:
That should be fine, but be sure to confirm with IPA/support team that this solution will be sufficient for them, and there will be no automatic upgrade for existing instances.